### PR TITLE
[JENKINS-49235] allow enable/disable credentials tracking

### PR DESCRIPTION
**Pull request by kuisathaverat**
[JENKINS-49235](https://issues.jenkins-ci.org/browse/JENKINS-49235)

* Property `-Dhudson.plugins.sshslaves.SSHLauncher.trackCredentials=true` to enable/disable credentials tracking, enable by default since [JENKINS-38832](https://issues.jenkins-ci.org/browse/JENKINS-38832)
* Show configuration on Agent log before start

### CODE REVIEW

**Comment by kuisathaverat** [comment]
Even do the Andy Taylor's proposal is approved, I think that it is a good think to be able to disable it, indeed if the first implementation has been make on that form the user would have a workaround and not a critical issue. I like to add features but they should always have a way to disable it in case that causes an issue on a particular configuration.

**Comment by kuisathaverat** [comment]
you are right, this change it is not really useful because you have to disable the credentials tracking on each Agent, a tough work, so I will change it to a property to disable it on all SSH Agents.

File: doc/TROUBLESHOOTING.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+
+## Troubleshooting
+
+### Common Pitfalls
+#### Login profile files
+When the SSH slaves plugin connects to a slave, it does not run an interactive shell. 
+Instead it does the equivalent of your running `ssh slavehost command...` a few times,
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: Remove "your".

File: doc/TROUBLESHOOTING.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+
+## Troubleshooting
+
+### Common Pitfalls
+#### Login profile files
+When the SSH slaves plugin connects to a slave, it does not run an interactive shell. 
+Instead it does the equivalent of your running `ssh slavehost command...` a few times, 
+eventually to run `ssh slavehost java -jar ...`. Exactly what happens on the slave as a result of this depends on the SSHD implementation,
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: "eventually to run" -> "eventually running"

File: doc/TROUBLESHOOTING.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+
+## Troubleshooting
+
+### Common Pitfalls
+#### Login profile files
+When the SSH slaves plugin connects to a slave, it does not run an interactive shell. 
+Instead it does the equivalent of your running `ssh slavehost command...` a few times, 
+eventually to run `ssh slavehost java -jar ...`. Exactly what happens on the slave as a result of this depends on the SSHD implementation, 
+but OpenSSH runs this with `bash -c command ...` (or whatever your login shell is.)
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: I'd suggest removing the "but" and separating this into two sentences -> "SSHD implementation. OpenSSH runs"

File: doc/TROUBLESHOOTING.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+This means some of your login profiles that set up your environment is not read by your shell. See [this post](http://stackoverflow.com/questions/216202/why-does-an-ssh-remote-command-get-fewer-environment-variables-then-when-run-man) for more details.
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: "is" -> "are" 

("Some" indicates a likelihood of more than one, so it uses the plural form.)

File: doc/TROUBLESHOOTING.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+Example: For loading `~/.bash_profile` on macOS Slaves set **Prefix Start Slave Command** to `source ~/.bash_profile && ` (intended whitespace at the end).
+Make sure to reconnect the slave after changing the slaves commands.
+
+#### Overall recommendations:
+* It is recommended to use JDK nearest and in the same major version of Jenkins instance and Agents
+* It is recommended to tune the TCP stack on of Jenkins instance and Agents
+    * On Linux http://tldp.org/HOWTO/TCP-Keepalive-HOWTO/usingkeepalive.html    
+    * On Windows https://blogs.technet.microsoft.com/nettracer/2010/06/03/things-that-you-may-want-to-know-about-tcp-keepalives/
+    * On Mac https://www.gnugk.org/keepalive.html
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: Suggested simplifications:

* Use a JDK version in the same major version as the Jenkins instance and Agents and preferably a close minor version.
* Tune the TCP stack on of Jenkins instance and Agents
    * [Linux](http://tldp.org/HOWTO/TCP-Keepalive-HOWTO/usingkeepalive.html)
    * [Windows](https://blogs.technet.microsoft.com/nettracer/2010/06/03/things-that-you-may-want-to-know-about-tcp-keepalives/)
    * [Mac](https://www.gnugk.org/keepalive.html)

File: doc/TROUBLESHOOTING.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+* You should check for hs_err_pid error files in the root fs of the agent http://www.oracle.com/technetwork/java/javase/felog-138657.html#gbwcy
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: "fs" => "filesystem"

File: doc/TROUBLESHOOTING.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+locked hudson.model.Fingerprint@40e3a4f1
+at hudson.BulkChange.commit(BulkChange.java:98)
+at com.cloudbees.plugins.credentials.CredentialsProvider.trackAll(CredentialsProvider.java:1533)
+at com.cloudbees.plugins.credentials.CredentialsProvider.track(CredentialsProvider.java:1478)
+at hudson.plugins.sshslaves.SSHLauncher.launch(SSHLauncher.java:856)
+locked hudson.plugins.sshslaves.SSHLauncher@57dc4a8a
+...
+```
+
+You will need to disable the credentials tracking by setting the property `-Dhudson.plugins.sshslaves.SSHLauncher.trackCredentials=true`
```
Comment (jeffret-b) [inline doc/TROUBLESHOOTING.md]: Maybe change "You will need to disable" to something like "You may want to disable". I'm not sure, but the current wording feels a little strong, particularly if they wanted credential tracking.

**Comment by kuisathaverat** [comment]
>are you looking to do a release of the plugin to the main update site soon?

I want to documented some important changes, and make a configuration guide from scratch, I want to release it on no more than two weeks.

* Jenkins 2.73.3 is required
* The slave.jar file is changed to remoting.jar [JENKINS-44111](https://issues.jenkins-ci.org/browse/JENKINS-44111)
* A JDK/JRE 8 installed is required to run the remoting.jar process [JENKINS-47586](https://issues.jenkins-ci.org/browse/JENKINS-47586)
* Remoting workdir is supported by default
* Timeout, retries, and time between retries have a default values [JENKINS-52613](https://issues.jenkins-ci.org/browse/JENKINS-52613)
* Support for Win32-OpenSSH [JENKINS-37152](https://issues.jenkins-ci.org/browse/JENKINS-37152)

>would you recommend one of those over the other?

I recommend to use the master version, it will be the released version.

**Comment by jglick** [comment]
Given that https://github.com/jenkinsci/credentials-plugin/pull/103 has been merged, can this be reverted? Or at least the documentation should be edited to clarify that users would no longer normally run into this issue.

**Comment by kuisathaverat** [comment]
I think we can keep it, the documentation related is in the troubleshooting guide in the section [Threads stuck at CredentialsProvider.trackAll](https://github.com/jenkinsci/ssh-slaves-plugin/blob/master/doc/TROUBLESHOOTING.md#threads-stuck-at-credentialsprovidertrackall) the title point to a specific problem so I think is fine.

**Pull request merged**